### DISCUSSION TITLE
Suggestion: nicegui + brython = full stack python development

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1904] [Ideas / Feature Requests]
Interested to hear feedback on this proposal.  @rodja @falkoschindler would this be of any interest to you?

# Question
How do people currently do full stack web development in python?  Nicegui is great, it handles a lot.  Backend server, frontend gui, communication between the two.

But sometimes you need to run code on the client side, for various reasons:
- Some resources are only accessible in the browser.  
- Some computation is best done on the client.  
- Sometimes network lag or bandwidth is an issue.

It's great that nicegui lets you run javascript on the client side.  But this has limitations and gotchas.  Javascript adds more cognitive load.  It would be great if client-side code could also be written in python...

# Potential Solution
[Brython](https://brython.info/) is python in the browser.  A full feature-complete (comparable to cpython) python implementation with standard lib, implemented in javascript.  Just load brython.js in your page and you can run python scripts like this:
```html
<script type="text/python">
print('hello world')
</script>
```

Brython is quite mature, has good docs and good performance.  It was recently added to the [DroidScript](https://droidscript.org/) mobile app IDE with almost 2M downloads.  Brython uses a BSD license compatible with Nicegui.

I've used brython for 5 years and been very happy with it.  The things it lacks - gui toolkit, backend server, communication - are exactly the things nicegui provides.  They would work together very well.

# Ways Forward
I plan to explore using nicegui and brython together.  My initial thought is this:
- add brython.js to my nicegui app with `ui.add_html_head ('<script src="brython.js">')`
- add brython scripts (i.e. client-side python code) with `ui.element ('<script type="text/python"> code here </script>')`
- invoke brython code when needed.  options : 
  1. [register brython function with window](https://brython.info/static_doc/3.12/en/jsobjects.html#brython_from_js), invoke from javascript
  2. bind [brython handlers](https://brython.info/static_doc/3.12/en/events.html) to certain events / ui elements and not handle them in nicegui
  3. [invoke the brython interpreter](https://brython.info/static_doc/3.12/en/options.html) to run a particular script when needed

At least one of these approaches should work.  In theory, it should work fine and not interefere with nicegui / quasar / fastapi at all.  But until I test it, I won't know the effects for sure.

# Closer Integration
The above should work without any changes to nicegui.  However, better integration could make it even easier to use nicegui and brython together.  Imagine writing code like this:
```py
def run_on_client():
    """brython code to update time in label every second"""

    from browser import document
    label = document ['clock']
    label.set_text(f'{datetime.now():%X}')
    browser.set_timeout(run_on_client, 1000)  # call again in one second

ui.label().props('id=clock')
ui.button('Go', on_click_client = run_on_client)
ui.run()
```

Nicegui can already achieve the same effect; I took the example from the [timer page](https://nicegui.io/documentation/timer).  But in current nicegui, each update requires communication between server and client, if my understanding of nicegui architecture is correct.

With brython example above, all toggles would happen on the client side with no server communication.  A little behind-the-scenes magic would be required to make this happen.  Nicegui could bundle and send the run_on_client code once to brython for execution, either as plain text or AST.

This simple example shows how nicegui + brython could work together to reduce server load and improve responsiveness in certain situations.

Testing brython + nicegui is on my todo list, but I probably won't get to it for awhile due to higher priorities.  I will try it eventually though.

**Comment by dridk** [comment]
Pyodide should be a better solution

https://pyodide.org/en/stable/

**Reply by ed2050** [reply]
Based on what?  Pyodide is web assembly.  From [what I can see](https://pyodide.org/en/stable/project/roadmap.html
), it's less mature, slower to load, larger download, doesn't support synchronous io, supports web workers poorly, etc.  Not so great.

But the big issues are not pyodide itself but web assembly limitations: 
- limited interaction with the DOM, making it unsuitable for interactive user interfaces
- lack of garbage collection
- security issues.  web assembly engines are much less mature than js engines.

Pyodide can access the DOM but everything [goes through the js module](https://pyodide.org/en/0.16.1/using_pyodide_from_javascript.html).  This has a [significant performance penalty](https://pyodide.org/en/0.16.1/type_conversions.html).

It seems that brython is the better solution, unless I'm missing something.

**Comment by rodja** [comment]
@ed2050 thanks for bringing up this interesting combination. We have not tried Brython ourself but it sounds interesting. I'm currently not sure if it can be nicely integrated into this project or would be better off as an separate project on top of NiceGUI. Maybe instead of adding `on_click_client` and other specialized handlers, a decorator `@browser` could mark functions which are executed in the browser. 

Anyway we would need to figure out how Brython can find the objects in the DOM. Your code `label = document ['clock']` won't work because NiceGUI uses generated ids which can not be altered. Also I wonder how we keep frontend and backend in sync. One of the things which makes NiceGUI so strong is that you have all the state in the backend...

**Comment by falkoschindler** [comment]
Thanks for the suggestion, @ed2050!
Yes, running Python on the client would be nice.

> But in current nicegui, each update requires communication between server and client

Yes, unless you implement your own Vue component that runs on the client, like shown in our [Custom Vue component example](https://github.com/zauberzeug/nicegui/tree/main/examples/custom_vue_component). Of course it would be great to support Python for such components.

Regarding your example, I have some concerns:

- The IDE, type checkers and linters will complain about it quite a lot. Since `browser` probably can't be imported on the Desktop, the remaining code block cannot be understood.
- The `label` must of some Brython type representing an HTML element. But it isn't a NiceGUI element, because there isn't necessarily a one-to-one mapping between HTML and NiceGUI (see, e.g., `ui.input`) and Brython doesn't know about NiceGUI. Therefore `set_text` must be a Brython method, not to confuse with `ui.label`'s `set_text()`.
- The Brython code shares variable scope with the remaining NiceGUI app, which seems very error-prone. As soon as you access Brython values from NiceGUI or the other way around, it will fail when run on the client.
- To avoid confusion, I think it would be better to move Brython code into separate files, like we do with Vue components. The integration wouldn't be as seamless, but I think there has to be a clear separation anyway.

To move forward I guess we need to re-think the potential application. I would start with the "Custom Vue component" example and try to replace counter.js with Brython code ("wishful programming", doesn't have to execute yet). Or maybe there is another way to integrate frontend and backend which I'm not seeing right now.

**Comment by ed2050** [comment]
Thanks @rodja @falkoschindler.  All good points.  My takeaway is that you find the concept appealing on some level, but there would be several issues to work through before considering it a viable option.  Which is totally fair.  That's what I expected.

I understand the state concerns between client and server.  Is it any different than maintaining state between python and javascript?  However you handle state in js could possibly be applied to brython.

I'm sure we could find ways to address the other issues.

# Demo code
You're right, the code example wouldn't work as currently written.  It's just a conceptual demo.

As currently constructed, brython could interact with the nicegui object using the normal DOM node interface.  Brython `label = document ['clock']` retrieves the node with id=clock in brython.  That's not the same as retrieving the nicegui object, but for this example it's ok.  The brython way to update the label text would be `label.text = f'{datetime.now():%X}'` instead of set_text.  That part could work today.

For brython code to have the nicegui object interface, something would have to tell brython this DOM node is actually a nicegui object.  That something could be a custom VUE component, or an interface layer, or something else.  And there could be better ways to retrieve the object than by DOM id (maybe it's passed as a function argument).  Good things to think about.

I'll come back to this when I have more time to experiment.  Thanks for being open to the possibility.

**Comment by evnchn** [comment]
With reference to: 

https://brython.info/static_doc/3.13/en/intro.html
https://github.com/zauberzeug/nicegui/discussions/637#discussioncomment-5856190
https://github.com/zauberzeug/nicegui/pull/4552#issuecomment-2773919564
https://github.com/miguelgrinberg/microdot (v1, since v2 [No more synchronous web server](https://microdot.readthedocs.io/en/stable/migrating.html#no-more-synchronous-web-server))

Apparently, I was able to get some small part of microdot working on Brython on the browser by the following hacky approach: 

```html
<!doctype html>
<html>

<head>
<meta charset="utf-8">
<script type="text/javascript" src="brython.js"></script>
<script type="text/javascript" src="brython_stdlib.js"></script>
</head>

<body>
<script type="text/python">
from microdot import Microdot

from browser import document

document <= 'Test'

app = Microdot()

@app.route('/')
def index():
    return 'Hello, world!'

print(app.url_map)

document <= app.url_map[0][2]() # forcibly calling the right page function, since app.run() won't work most probably
</script>
</body>

</html>
```

But, this goes to show that, hey Python on the browser, it could work. 

---

If we want to make this work for NiceGUI (think: `from nicegui import ui` in the browser), the following massive changes has to be made: 

* Remove / not load the dependencies that won't run on the browser, such as FastAPI, SSL, any native C libraries, etc...
* Either give up async page function, or use [`browser.aio`](https://www.brython.info/static_doc/3.13/en/browser.aio.html)
* Rethink the Socket.IO with some other connection means between the Brython (the page definition) and the JavaScript (`nicegui.js`)

No time to experiment with that lately, but just want to add another idea to the thread.
